Update use of coredis to use 3.x API

alisaifee opened the pull request:
# Description
There are a bunch of breaking api changes between coredis 2.x (backward compatible with aredis) and coredis 3.x. I won't be maintaining the 2.x branch for much longer and have updated the limits library to use 3.x as well.

## References
- [Major changes in 3.x](https://coredis.readthedocs.io/en/stable/release_notes.html#v3-0-0)

## Misc Notes
- Python 3.7 support is dropped

alisaifee committed:
Update use of coredis to use 3.x API

alisaifee committed:
Replace StrictRedis->Redis

alisaifee commented:
Looks like a lot of the tests aren't happy with this change. I'll close this for now but happy to revive it if it's of interest to update the dependency.

zoltan-fedor commented:
Hi @alisaifee,
Thanks for the PR!
What do you mean by "a lot of the tests aren't happy with this change"?
The tests were fine, see https://github.com/zoltan-fedor/falcon-caching/pull/7/checks, only the codecove was shopping some drop in coverage in `falcon_caching/async_backends/redis.py`, but in total the codecov change was still positive, so all is good.
Let's revive this and get that dependency updated.
Thanks again!

alisaifee commented:
Thanks for following up @zoltan-fedor ! 
I might have gotten myself a bit confused with the tests as when I ran a subset locally I was getting some errors related to unlinking and in CI (for example [here](https://github.com/zoltan-fedor/falcon-caching/runs/6038772436?check_suite_focus=true) I see a lot of async+redis failures). Happy to work through them though.

All in all - the backward incompatibilities boil down to calls to `delete, unlink and exists` which used to take variadic arguments but now expect a list.

zoltan-fedor commented:
Hi @alisaifee,
Actually you were right, it was me who got confused. I have only recently switched to using GitHub Actions for CI and it seems for some reason something is wrong, as GitHub Actions show the tests as successful, while when you drill down you see them lots of failures.
It seems most are related to attribute errors:
```
AttributeError: 'Redis' object has no attribute 'get_many'
AttributeError: 'Redis' object has no attribute 'clear'
```
Did these attributes got removed with 3.0 of `coredis`?

alisaifee commented:
> ```
> AttributeError: 'Redis' object has no attribute 'get_many'
> AttributeError: 'Redis' object has no attribute 'clear'
> ```
> Did these attributes got removed with 3.0 of `coredis`?

That's the curious part - they were never there! `get_many` & `clear` appear to be methods that should be expected from `falcon_caching.async_backends.base.BaseCache` - not `coredis.Redis`. My guess is that somehow the tests are accessing the underlying storage instead of the wrapper cache class?

alisaifee committed:
Ensure coredis.Redis doesn't override the global namespace

alisaifee commented:
@zoltan-fedor turned out to be something silly - as of 3.x I've renamed `StrictRedis/StrictRedisCluster` to just `Redis/RedisCluster` and renaming that ended up clobbering the `falcon_caching.async_backends.Redis` class name. Fixed in e531ff50309640b2ce116e4f19f442f626630315

zoltan-fedor commented:
Hi @alisaifee,
Hah, name conflict. That explains why those missing argument errors were popping up all around. Thanks for fixing it!
I am going to merge it and make a new release with this included.
Thanks

alisaifee commented:
Nice!
Hopefully the dropped 3.7 support doesn't affect anyone!

zoltan-fedor commented:
Nah. I have introduced async support recently - I doubt that anybody would be using Python 3.7 on any recently developed project, so we are likely good.
